[v15] Bump Node.js (v22) and Electron (v35), remove manual install of corepack from build scripts

ravicious opened the pull request:
Backport #53610.

Tag build: https://github.com/gravitational/teleport.e/actions/runs/14475193186

Fewer changes in this backport since the corepack changes don't need to be backported.

dependabot[bot] committed:
Bump Node.js (v22) and Electron (v35), remove manual install of corepack from build scripts

* Bump the electron group with 3 updates

Bumps the electron group with 3 updates: [electron](https://github.com/electron/electron), [electron-builder](https://github.com/electron-userland/electron-builder/tree/HEAD/packages/electron-builder) and [electron-vite](https://github.com/alex8088/electron-vite).

Updates `electron` from 34.3.0 to 35.1.2
- [Release notes](https://github.com/electron/electron/releases)
- [Changelog](https://github.com/electron/electron/blob/main/docs/breaking-changes.md)
- [Commits](https://github.com/electron/electron/compare/v34.3.0...v35.1.2)

Updates `electron-builder` from 25.1.8 to 26.0.12
- [Release notes](https://github.com/electron-userland/electron-builder/releases)
- [Changelog](https://github.com/electron-userland/electron-builder/blob/master/packages/electron-builder/CHANGELOG.md)
- [Commits](https://github.com/electron-userland/electron-builder/commits/v26.0.12/packages/electron-builder)

Updates `electron-vite` from 3.0.0 to 3.1.0
- [Release notes](https://github.com/alex8088/electron-vite/releases)
- [Changelog](https://github.com/alex8088/electron-vite/blob/master/CHANGELOG.md)
- [Commits](https://github.com/alex8088/electron-vite/compare/v3.0.0...v3.1.0)

---
updated-dependencies:
- dependency-name: electron
  dependency-version: 35.1.2
  dependency-type: direct:development
  update-type: version-update:semver-major
  dependency-group: electron
- dependency-name: electron-builder
  dependency-version: 26.0.12
  dependency-type: direct:development
  update-type: version-update:semver-major
  dependency-group: electron
- dependency-name: electron-vite
  dependency-version: 3.1.0
  dependency-type: direct:development
  update-type: version-update:semver-minor
  dependency-group: electron
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump @electron/notarize from 2.5.0 to 3.0.1

Bumps [@electron/notarize](https://github.com/electron/notarize) from 2.5.0 to 3.0.1.
- [Release notes](https://github.com/electron/notarize/releases)
- [Changelog](https://github.com/electron/notarize/blob/main/.releaserc.json)
- [Commits](https://github.com/electron/notarize/compare/v2.5.0...v3.0.1)

---
updated-dependencies:
- dependency-name: "@electron/notarize"
  dependency-version: 3.0.1
  dependency-type: direct:development
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump node-pty from 1.1.0-beta14 to 1.1.0-beta33

Bumps [node-pty](https://github.com/microsoft/node-pty) from 1.1.0-beta14 to 1.1.0-beta33.
- [Release notes](https://github.com/microsoft/node-pty/releases)
- [Commits](https://github.com/microsoft/node-pty/commits/v1.1.0-beta33)

---
updated-dependencies:
- dependency-name: node-pty
  dependency-version: 1.1.0-beta33
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

* Bump electron to 35.1.4

* Remove override for electron-vite

* Remove unnecessary node_gyp config option from .npmrc

It's enough to have a specific version of node-gyp in package.json and
that will cause node-pty to use that over the version shipped with pnpm.

* Bump Node.js version to 22.14.0

* Remove manual installation of corepack@0.31.0 and new integrity keys

* Fix electron-builder deps

This is done by removing electron-builder from package.json, running
pnpm install and then adding it back again and running pnpm install.

* Do not install missing peers to solve problem with electron-builder-squirrel-windows

* Move Windows signing configuration to `signtoolOptions`

* Update @types/node to v22

* Bring back node_gyp config option in .npmrc

This reverts commit b565095500eb7c8da7b64b54656f9e6a13526705.

Turns out it's actually needed, as on Windows pnpm uses 11.1.0 without
this.

* Update node-gyp to 11.2.0

* Update linux templates

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Rafał Cieślak <rafal.cieslak@goteleport.com>
Co-authored-by: Grzegorz Zdunek <grzegorz.zdunek@goteleport.com>

ravicious committed:
Dedupe packages

ravicious commented on web/packages/build/package.json:
```diff
@@ -62,7 +62,7 @@
     "cross-env": "5.0.5",
     "cross-spawn": "6.0.5",
-    "electron-builder": "^25.1.8",
+    "electron-builder": "^26.0.12",
```
https://github.com/gravitational/teleport.e/actions/runs/14450639036/job/40523441669#step:18:169

```
$ electron-builder build --config electron-builder-config.js --publish never -c.extraMetadata.name=teleport-connect -c.extraMetadata.version=15.4.31-dev.ravicious.2 --universal
  • electron-builder  version=26.0.12 os=22.6.0
  • loaded configuration  file=/Users/runner/work/teleport.e/teleport.e/web/packages/teleterm/electron-builder-config.js
  • executing @electron/rebuild  electronVersion=35.1.4 arch=x64 buildFromSource=false appDir=./
  • installing native dependencies  arch=x64
  • preparing       moduleName=node-pty arch=x64
  ⨯ Could not detect abi for version 35.1.4 and runtime electron.  Updating "node-abi" might help solve this issue if it is a new release of electron  failedTask=build stackTrace=Error: Could not detect abi for version 35.1.4 and runtime electron.  Updating "node-abi" might help solve this issue if it is a new release of electron
    at Object.getAbi (/Users/runner/work/teleport.e/teleport.e/node_modules/node-abi/index.js:39:9)
    at get ABI (/Users/runner/work/teleport.e/teleport.e/node_modules/@electron/rebuild/lib/rebuild.js:89:39)
    at PrebuildInstall.prebuiltModuleExists (/Users/runner/work/teleport.e/teleport.e/node_modules/@electron/rebuild/lib/module-type/prebuild-install.js:61:179)
    at ModuleRebuilder.prebuildInstallNativeModuleExists (/Users/runner/work/teleport.e/teleport.e/node_modules/@electron/rebuild/lib/module-rebuilder.js:70:37)
    at Rebuilder.rebuildModuleAt (/Users/runner/work/teleport.e/teleport.e/node_modules/@electron/rebuild/lib/rebuild.js:143:35)
    at async Rebuilder.rebuild (/Users/runner/work/teleport.e/teleport.e/node_modules/@electron/rebuild/lib/rebuild.js:115:17)
error Command failed with exit code 1.
```

ravicious commented:
I'll check it later and bump node-abi.

ravicious committed:
Remove engines.node from package.json

In yarn it's a hard error instead of a warning.

ravicious commented:
> I'm getting changes in the lockfile after installing the packages. Perhaps something wasn't committed?

I forgot that after `npx yarn-deduplicate yarn.lock` we have to run `yarn install` once again to get rid of those changes. 😭

ravicious committed:
Re-run yarn install after dedupe

ravicious committed:
Bump node-abi to 3.74.0

It matches the version on master.

ravicious committed:
Merge branch 'branch/v15' into r7s/v15/backport-53610-node-v22

ravicious commented on web/packages/teleterm/package.json:
```diff
@@ -35,7 +35,7 @@
   },
   "devDependencies": {
-    "@electron/notarize": "^2.4.0",
+    "@electron/notarize": "^3.0.1",
```
I have to drop this to 2.5.0. 3.0.0 removed support for Node.js v20 and specifiecs `engines.node`. Yarn does a hard fail if `engines.node` doesn't match what's on the system.  Our CI is still on v20, so this won't go through until we add `--ignore-engines` to `yarn install`.

Instead of doing that, I'll just downgrade this package. It should still work and Connect v15 is going to be EOL soon anyway.

gzdunek replied:
Makes sense, let's downgrade the package.

ravicious committed:
Downgrade @electron/notarize to 2.5.0

ravicious commented:
I verified that Connect and Web UI from the tag build work correctly.

ravicious committed:
Add any to cloneAbortSignal

v15 is on an old TypeScript version which still thinks that `any` is
an instance method instead of a static method, so we have to add it there.

https://github.com/gravitational/teleport/pull/52683#discussion_r1977405260
